### PR TITLE
Wire up Top aggregate function

### PR DESCRIPTION
**Pull request by corylanou**
## Issue
This addresses https://github.com/influxdb/influxdb/issues/1821
Also of interest (but not solved for in this PR) is Bottom: https://github.com/influxdb/influxdb/issues/1820

## Aggregate Validations
We were doing aggregate validations in two separate spots.  This PR consolidates all of it to fail faster in the parser validation, not in the function mapping validation.

### CODE REVIEW

File: influxql/ast.go [hunk 1]
```diff
@@ -1036,27 +1036,76 @@ func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
...
+
+	allowMixedAggregates := func() error {
+		if numAggregates != 0 && numAggregates != len(s.Fields) {
+			return fmt.Errorf("mixing aggregate and non-aggregate queries is not supported")
+		}
+		return nil
 	}
 
-	// Secondly, determine if specific calls have at least one and only one argument
+	// Secondly, determine if specific calls have the correct number of arguments
```
Comment (corylanou) [inline influxql/ast.go]: This still needs more refactoring.  I wanted to at least get it all in the same spot for now.

File: helpers/slices/strings.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package slices
```
Comment (corylanou) [inline helpers/slices/strings.go]: I'm not a fan of `utility` packages, so if anyone has a better suggestion on where to put this/name it fire away.

File: tsdb/executor.go [hunk 3]
```diff
@@ -475,6 +471,12 @@ func (e *SelectExecutor) executeAggregate(out chan *influxql.Row) {
 			}
 		}
 
+		// Perform top/bottom unwraps
+		values, err = e.processTopBottom(values, columnNames)
```
Comment (corylanou) [inline tsdb/executor.go]: This will likely change to `processAggregates` in the future and branch from there, as we are going to add support for `MAX`, `MIN`, `FIRST`, `LAST` to also have additional tags/fields returned.

File: tsdb/mapper.go [hunk 4]
```diff
@@ -313,7 +313,7 @@ func (lm *SelectMapper) nextChunkRaw() (interface{}, error) {
 		}
 		cursor := lm.cursors[lm.currCursorIndex]
 
-		k, v, t := cursor.Next(lm.queryTMin, lm.queryTMax, lm.selectFields, lm.whereFields)
+		k, v := cursor.Next(lm.queryTMin, lm.queryTMax, lm.selectFields, lm.whereFields)
```
Comment (corylanou) [inline tsdb/mapper.go]: With the addition of `Tags()` to the iterator interface, we no longer needed to expand this signature and can ask for tags only when necessary.  This is an optimization and one that @otoolep was concerned about when we originally started returning tags with each `Next`.

File: helpers/slices/strings.go (hunk 2 above)
Comment (jwilder) [inline helpers/slices/strings.go]: I've seen some projects us `pkg` as the root for utility packages that are usable outside of the project.  See https://github.com/docker/docker/tree/master/pkg,  https://github.com/coreos/etcd/tree/master/pkg and https://github.com/kubernetes/kubernetes/tree/master/pkg.  They sometimes put a `README.md` in there as well.  I kind of like `pkg` as opposed to `helpers`.

File: influxql/functions.go [hunk 5]
```diff
@@ -1035,6 +1082,412 @@ func ReduceLast(values []interface{}) interface{} {
...
+	}
+
+	for k, v := itr.Next(); k != -1; k, v = itr.Next() {
+		t := k
+		if bt := itr.TMin(); bt > -1 {
+			t = bt
+		}
+		callArgs := c.Fields()
+		tags := itr.Tags()
+		// TODO send fields in
```
Comment (jwilder) [inline influxql/functions.go]: Did this get done?

Comment (corylanou) [inline influxql/functions.go]: No, we are doing it in a future PR, I left it in as a future TODO.  Is there a better way to note that?

File: tsdb/executor.go [hunk 6]
```diff
@@ -622,6 +624,61 @@ func (e *SelectExecutor) close() {
...
+
+	// Check if we have a group by, if not, rewrite the entire result by flattening it out
+	//if len(e.stmt.Dimensions) == 0 {
+	for _, vals := range results {
+		// start at 1 because the first value is always time
+		for j := 1; j < len(vals); j++ {
+			tMin := vals[0].(time.Time)
+			switch v := vals[j].(type) {
+			case influxql.PositionPoints:
+				for _, p := range v {
```
Comment (jwilder) [inline tsdb/executor.go]: Quite a bit of indenting here.  Maybe move to a func and do something like `values = e.toQueryResult(v)`?

File: influxql/functions.go (hunk 5 above)
Comment (jwilder) [inline influxql/functions.go]: I don't know what the TODO actually means.  Maybe elaborate a bit more on what is missing and needs to be done?

**Comment by jwilder** [comment]
A few small things but LGTM. :+1:.

File: CHANGELOG.md [hunk 7]
```diff
@@ -11,6 +11,8 @@ With this release InfluxDB is moving to Go 1.5.
 - [#3876](https://github.com/influxdb/influxdb/pull/3876): Allow the following syntax in CQs: INTO "1hPolicy".:MEASUREMENT
 - [#3975](https://github.com/influxdb/influxdb/pull/3975): Add shard copy service
 - [#3986](https://github.com/influxdb/influxdb/pull/3986): Support sorting by time desc
+- [#3930](https://github.com/influxdb/influxdb/pull/3930): Wire up TOP aggregate function
```
Comment (otoolep) [inline CHANGELOG.md]: Duplicating this in the CHANGELOG looks wrong to me. I think it's better to keep the first line, and then add "fixes [#1821](https://github.com/influxdb/influxdb/issues/1821)" to the line.

File: cmd/influxd/run/server_test.go [hunk 8]
```diff
@@ -993,7 +993,7 @@ func TestServer_Query_Count(t *testing.T) {
 		&Query{
 			name:    "selecting count(*) should error",
 			command: `SELECT count(*) FROM db0.rp0.cpu`,
-			exp:     `{"results":[{"error":"expected field argument in count()"}]}`,
+			exp:     `{"error":"error parsing query: expected field argument in count()"}`,
```
Comment (otoolep) [inline cmd/influxd/run/server_test.go]: Why is this changing?

File: influxql/ast.go [hunk 9]
```diff
@@ -857,6 +859,48 @@ func (s *SelectStatement) RewriteDistinct() {
 	}
 }
 
+// ColumnNames will walk all fields and functions and return the appropriate field names for the select statement
+// while maintaining sort order of the field names
```
Comment (otoolep) [inline influxql/ast.go]: Just so I am clear, it will maintain the order of whatever is in `s.Fields`? 'cos I see no sorting going on in this function.

File: cmd/influxd/run/server_test.go (hunk 8 above)
Comment (corylanou) [inline cmd/influxd/run/server_test.go]: The error is now caught upstream further, so it isn't being wrapped by an error return again, or in this specific case, a results object.

File: influxql/ast.go (hunk 9 above)
Comment (corylanou) [inline influxql/ast.go]: yes, the order that was specified in the query itself.

File: CHANGELOG.md (hunk 7 above)
Comment (corylanou) [inline CHANGELOG.md]: agreed, I'll make the change.

File: influxql/ast.go (hunk 9 above)
Comment (otoolep) [inline influxql/ast.go]: OK, minor but I would removed the word "sort" from the comment. All the code is doing is keeping the existing order.

File: influxql/ast.go [hunk 10]
```diff
@@ -1029,40 +1073,91 @@ func (s *SelectStatement) validateFields() error {
 	return nil
 }
 
+func (s *SelectStatement) allowMixedAggregates(numAggregates int) error {
```
Comment (otoolep) [inline influxql/ast.go]: FWIW, this feels more like it should be a boolean return.

Comment (otoolep) [inline influxql/ast.go]: So I understand this.

`top(), value` is allowed.
`top(), mean(), value()` is not.
`top(), mean()` is allowed.

Is that correct?

File: influxql/ast.go [hunk 11]
```diff
@@ -1029,40 +1073,91 @@ func (s *SelectStatement) validateFields() error {
...
+					return fmt.Errorf("invalid number of arguments for %s, expected at least %d, got %d", expr.Name, exp, got)
+				}
+				if len(expr.Args) > 1 {
+					callLimit, ok := expr.Args[len(expr.Args)-1].(*NumberLiteral)
+					if !ok {
+						return fmt.Errorf("expected integer as last argument in %s(), found %s", expr.Name, expr.Args[len(expr.Args)-1])
+					}
+					// Check if they asked for a limit smaller than what they passed into the call
+					if int64(callLimit.Val) > int64(s.Limit) && s.Limit != 0 {
+						return fmt.Errorf("limit (%d) in %s function can not be larger than the LIMIT (%d) in the select statement", int64(callLimit.Val), expr.Name, int64(s.Limit))
```
Comment (otoolep) [inline influxql/ast.go]: Nit "can not" -> "cannot".

File: influxql/ast.go (hunk 9 above)
Comment (corylanou) [inline influxql/ast.go]: agreed, will change.

File: influxql/ast.go (hunk 11 above)
Comment (otoolep) [inline influxql/ast.go]: Actually, checking online, can not is also OK. Guess I am so used to seeing it as one word. :-)

File: influxql/ast.go [hunk 12]
```diff
@@ -2241,6 +2336,33 @@ func (c *Call) String() string {
 	return fmt.Sprintf("%s(%s)", c.Name, strings.Join(str, ", "))
 }
 
+// Fields will extract any field names from the call.  Only specific calls support this.
+func (c *Call) Fields() []string {
+	switch c.Name {
+	case "top", "bottom":
+		// maintain the sort order
```
Comment (otoolep) [inline influxql/ast.go]: Again, unless you're actually sorting I would just say "maintain existing order".

File: influxql/ast.go [hunk 13]
```diff
@@ -2241,6 +2336,33 @@ func (c *Call) String() string {
...
+		keyMap := make(map[string]struct{})
+		keys := []string{}
+		for i, a := range c.Args {
+			if i == 0 {
+				// special case, first argument is always the name of the function regarldess of the field name
```
Comment (otoolep) [inline influxql/ast.go]: Nit: regarldess -> regardless

File: influxql/ast_test.go [hunk 14]
```diff
@@ -451,7 +451,7 @@ func TestSelectStatement_IsRawQuerySet(t *testing.T) {
 			isRaw: false,
 		},
 		{
-			stmt:  "select mean(*) from foo group by *",
```
Comment (otoolep) [inline influxql/ast_test.go]: Why this change?

File: influxql/ast.go (hunk 10 above)
Comment (corylanou) [inline influxql/ast.go]: Sort of.  So clearly the name is horrible.  I'll think on a better one.

In theory `TOP`, `BOTTOM`, `MIN`, `MAX`, `FIRST`, and `LAST` can all be called with additional fields/tags in the query, but other aggregates, like `MEAN` can ONLY be called by themselves, or potentially with other aggregates (I'm not sure about this last part though).

Let me think about how to make this better.

File: influxql/ast.go (hunk 11 above)
Comment (corylanou) [inline influxql/ast.go]: Fair enough.  I won't make fun of when you write `favour` either :-).

File: influxql/ast.go (hunk 12 above)
Comment (corylanou) [inline influxql/ast.go]: agreed

File: influxql/ast_test.go (hunk 14 above)
Comment (corylanou) [inline influxql/ast_test.go]: `mean(*)` isn't currently valid in our system.  Most of the aggregates don't know how to handle a `*`  and require a field.  With the changes to the validator picking this up earlier now, it surfaced this in the test and had to be fixed.  In the future this restriction will be lifted.

**Comment by otoolep** [comment]
OK, took a look, my understand of the work required isn't super-deep, but we have decent testing around this so it should be OK.

+1, some minor feedback.

File: influxql/ast_test.go (hunk 14 above)
Comment (otoolep) [inline influxql/ast_test.go]: Cool.

**Comment by corylanou** [comment]
@otoolep addressed feedback.

**Comment by otoolep** [comment]
Great, the feedback was minor so merge away on green.

**Comment by Jhors2** [comment]
Is this feature also nestable with aggregates?  I.e. ```top(sum(value), 10) from cpu group by hostname```

**Comment by prawnsalad** [comment]
@corylanou Going from the comments on a few issues, should we be able to run `TOP()` on an aggregate? ie. `top(sum(value), 10) from cpu group by hostname`

Testing this on 0.9.4.2 which the changelog mentions it should be possible (I think) however it does accept a nested aggregate.

**Comment by corylanou** [comment]
You can not do a nested aggregate in top currently,
